src/Workspaces/Core/Portable/Indentation/IIndentationService.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Indentation
+{
+    /// <summary>
+    /// An indentation result represents where the indent should be placed.  It conveys this through
+    /// a pair of values.  A position in the existing document where the indent should be relative,
+    /// and the number of columns after that the indent should be placed at.  
+    /// 
+    /// This pairing provides flexibility to the implementor to compute the indentation results in
+    /// a variety of ways.  For example, one implementation may wish to express indentation of a 
+    /// newline as being four columns past the start of the first token on a previous line.  Another
+    /// may wish to simply express the indentation as an absolute amount from the start of the 
+    /// current line.  With this tuple, both forms can be expressed, and the implementor does not
+    /// have to convert from one to the other.
+    /// </summary>
+    internal struct IndentationResult
+    {
+        /// <summary>
+        /// The base position in the document that the indent should be relative to.  This position
+        /// can occur on any line (including the current line, or a previous line).
+        /// </summary>
+        public int BasePosition { get; }
+
+        /// <summary>
+        /// The number of columns the indent should be at relative to the BasePosition's column.
+        /// </summary>
+        public int Offset { get; }
+
+        public IndentationResult(int basePosition, int offset) : this()
+        {
+            this.BasePosition = basePosition;
+            this.Offset = offset;
+        }
+    }
+
+    internal interface IIndentationService : ILanguageService
+    {
+        /// <summary>
+        /// Determines the desired indentation of a given line.  May return <see langword="null"/> if the
+        /// <paramref name="document"/> does not want any sort of automatic indentation.  May also return
+        /// <see langword="null"/> if the line in question is not blank and thus indentation should
+        /// be deferred to the formatting command handler to handle.
+        /// </summary>
+        IndentationResult? GetDesiredIndentation(Document document, int lineNumber, CancellationToken cancellationToken);
```

Comment:
Note: i still hate this (and do not understand why the API id designed this way).  I would really like to make this non-nullable, and move anything that the "formatting command handler" would do into this logic.   It seems super bizarre that we have this strange split instead of making the formatting command handler *not* handle 'enter' and having that just be smart idnent, which goes through this.  

Unfortunately, this side of formatting/indentation is just an utter mess and really hard to detween.

Reply:
I'm going to try to make this change as a followup.  But it would be nice to get this in first to keep the size of the PRs smaller.

Reply:
@heejaechang Can you shed some light here on the design.  There seem to be three different systems here, which all are stateful and which try to infer what the other would be up to.

First, there is ISmartIndent.  This is the system the editor calls in on us to ask us "where should we be indented to".  This seems like the most natural location to place any indentation logic/computation.

However, we seem to have *two other* systems that are external to this.  First, there is the SmartTokenFormatHandler.  This handles enter, and sometimes processes where the enter shoudl go.  However, it's not *always* the case that enter is handled here.  Sometimes this instead says "will FormatCommandHandler handle enter?" and then defers to *that* system to compute where enter can go.

It's really not clear here why there is so much split in responsibilities in terms of computing this.  Nor is it clear when each is supposed to be figuring this out.  I'd really like to move this all down to the IIndentationService.  First, as far as i can tell, this should always succeed (so we can get rid of an unnecessary nullable).  We should always be able to figure out indentation and the API should reflect taht.  Second, it would be nice to have a *single* service that any downstream consumer can use to ask "what should the indentation be here".  Right now, afaict, that's (nearly) impossible to ask since you would have to ask three separate systems "are *you* responsible for figuring out indentation?  and if so, what do you think the indentation should be?"   This seems highly error prone and confusing.

Can you shed some light here on why the design is this way, and how we might be able to remove this complexity and just have a single standard mechanism for doing this?

Reply:
I dont know what it is doing. I don't know who wrote this ISynchronousIndentationService, IBlankLineIndentationService. or IndentationResult.

what I wrote, I believe, was someting return virtual position on the blank line for editor API. 

hmm.. looking at what we have now (http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/SmartIndent/IIndentationService.cs,a4f4522e8b8eb434,references)

not even sure why the API is async. I guess someone just extract out things from editor API but didn't properly refactor it? ah.. no one in roslyn implement that service...

okay, I remember this ShouldUseSmartTokenFormatterInsteadOfIndenter, but loos like it is like moved from its original context and copied to this place... anyway, this is something that see whether we can use formatting rule to figure out indentation or not. if it can, it should use the rules so that indentation doesn't need to re-implement all those logic to figure out what is right indentation from formattiong options such as case should be indented or not, or brace should be indented or not and etc. list of all options is in formattion option page.

this indenter and abstract indenter code seems have my initial code but looks like moved around. but not actually properly refactored but just moved around.

and smart token formatter is the thing that contains all the heuristic to figure out whether what span it should call formatting to or whether it should just move caret to right indentation and etc.

type char handling could be moved from formatting command handler to this, but looks like formatting command handler is done in a way it can handle anything that is not related to indentation, but smart token command handler is written to handle case where 2 can get mixed.

anyway, I am not sure why IEditorFormattingService is introduced and why everything is refactored this way. I remember code reviewing this and gave some feedbacks on refactor it differently but I guess my feedback wasn't taken.

hmmm smart token formatter used to handle all cases such as format range, and etc, not sure why ISmartTokenFormatter only do FormatToken and FormatRange is not part of it and etc.

I guess it is time to clean things up here.

Reply:
Note: a later PR removes this.